### PR TITLE
feat(a380x-fire): fire protection system

### PR DESCRIPTION
**Pull request by mjuhe**
## Summary of Changes

This introduces the fire protection system for the A380x (ATA26).

Introduced in this PR:
- Logic for fire detection for each engine, APU and MLG
- Fire test pushbutton now works
- Logic for squibs arming
- Logic for bottle discharging (note we still don't have a way to recharge the bottles!)
- APU fire on ground auto extinguishing
- APU emergency shut down on ground when a fire occurs
- Failures (engine fire, apu fire, mlg fire, fire loop detection failures)

## Screenshots (if necessary)

Screenshot 2024-05-03 134952

## References
In addition to all the usual documentation, a fun video: https://youtu.be/rvTNYmbes0E?si=fH0GC0ccUnXRpPI3

## Additional context

This PR is the first of two parts, as it only adds the fire detection and extinguishing of the engines, APU and MLG. The smoke detection of cargo and cabin has been left for a future PR.

Note the overhead xml will need addressing in a future PR as well. At the moment we still don't have a clickspot for the fire guards, so I've set them up to use the same clickspot as the fire buttons for now. Input events are left for future work as well (as I couldn't figure them out).

When triggering engine are not "connected" to the engines at all, so even if a fire is triggered the engine continues to run as normal. This is also left as future work.

Discord username (if different from GitHub): Commander Ga

## Testing instructions

As this is A380 only, it doesn't require QA. However, here are some suggestions for things to try later on: 
Press the fire test pushbutton on the overhead and observe all lights come on.
Turn on the APU on the ground, trigger an APU fire through the EFB failures. Observe the fire light come on, the extinguishing  the APU go into emergency shut down and the agent automatically discharge after a 10s delay.
Trigger failures of detection loops, observe that a fire is detected if both loops fail within 5 seconds.

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

### CODE REVIEW

**Review by 2hwk — APPROVED** [review]

File: fbw-a380x/src/wasm/systems/a380_systems/src/fire_and_smoke_protection.rs [hunk 1]
```diff
@@ -0,0 +1,1553 @@
...
+                        .iter()
+                        .any(|l| l.loop_has_failed(zone)))
+                || (self
+                    .fire_detection_loop
+                    .iter()
+                    .all(|l| l.loop_has_failed(zone))
+                    && self.interval_between_loop_failures[id] < Duration::from_secs(5)
+                    && zone != FireDetectionZone::Mlg);
+        }
+        fire_detected
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/fire_and_smoke_protection.rs:244]: I tried your suggestions but they don't seem to work (applies for the one below too). This way the array is not actually being modified in the loop and the function always returns [false; 6].

File: fbw-common/src/wasm/systems/systems/src/shared/mod.rs [hunk 2]
```diff
@@ -517,6 +517,7 @@ pub(crate) enum FwcFlightPhase {
 
 /// The delay logic gate delays the true result of a given expression by the given amount of time.
 /// False results are output immediately.
+#[derive(Clone, Copy)]
```
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-common/src/wasm/systems/systems/src/shared/mod.rs:520]: Good spot, no we don't! I might've added it there while testing something. Removed now.

File: fbw-a380x/src/wasm/systems/a380_systems/src/fire_and_smoke_protection.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gurgel100) [reply fbw-a380x/src/wasm/systems/a380_systems/src/fire_and_smoke_protection.rs:244]: Yes sorry here the fixed version:
```suggestion
        let mut fire_detected = [false; 6];
        for ((&zone, &interval_between_loop_failures), fire_detected) in self
            .fire_detection_zones
            .iter()
            .zip(&self.interval_between_loop_failures)
            .zip(&mut fire_detected)
        {
            *fire_detected = (self.fire_detection_loop[0]
                .fire_detected_in_loop(zone, fire_test_pb)
                && self.fire_detection_loop[1].fire_detected_in_loop(zone, fire_test_pb))
                || (self
                    .fire_detection_loop
                    .iter()
                    .any(|l| l.fire_detected_in_loop(zone, fire_test_pb))
                    && self
                        .fire_detection_loop
                        .iter()
                        .any(|l| l.loop_has_failed(zone)))
                || (self
                    .fire_detection_loop
                    .iter()
                    .all(|l| l.loop_has_failed(zone))
                    && interval_between_loop_failures < Duration::from_secs(5)
                    && zone != FireDetectionZone::Mlg);
        }
        fire_detected
```

File: fbw-a380x/src/wasm/systems/a380_systems/src/fire_and_smoke_protection.rs [hunk 3]
```diff
@@ -0,0 +1,1553 @@
...
+                .fire_detection_loop
+                .iter()
+                .all(|l| l.loop_has_failed(zone))
+            {
+                self.interval_between_loop_failures[id]
+            } else {
+                self.interval_between_loop_failures[id] + context.delta()
+            }
+        }
+        interval
```
[the remark this answers is not in the document]
Comment (Gurgel100) [reply fbw-a380x/src/wasm/systems/a380_systems/src/fire_and_smoke_protection.rs:266]: Here a fixed version:
```suggestion
        let mut interval = [Duration::ZERO; 6];
        for ((&zone, &interval_between_loop_failures), interval) in self
            .fire_detection_zones
            .iter()
            .zip(&self.interval_between_loop_failures)
            .zip(&mut interval)
        {
            *interval = if self
                .fire_detection_loop
                .iter()
                .all(|l| !l.loop_has_failed(zone))
            {
                Duration::ZERO
            } else if self
                .fire_detection_loop
                .iter()
                .all(|l| l.loop_has_failed(zone))
            {
                interval_between_loop_failures
            } else {
                interval_between_loop_failures + context.delta()
            }
        }
        interval
```

File: fbw-a380x/src/wasm/systems/a380_systems/src/fire_and_smoke_protection.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjuhe) [reply fbw-a380x/src/wasm/systems/a380_systems/src/fire_and_smoke_protection.rs]: Thanks! Works perfectly now.

**Review by Gurgel100 — APPROVED** [review]

**Pull request merged**